### PR TITLE
quic: test fast-fail if secrets not loaded when create quic connection

### PR DESCRIPTION
**Pull request by YaoZengzeng**
Commit Message: quic: test fast-fail if secrets not loaded when create quic connection  
Additional Description: N/A
Risk Level: Low
Testing:  N/A
Docs Changes: N/A

### CODE REVIEW

**Comment by RyanTheOptimist** [comment]
Do you have any context for this change?
Please make sure we have a new unit/integration test.
It looks like there's a build problem at the moment which should have been fixed by https://github.com/envoyproxy/envoy/pull/18708. I recommend syncing and trying again?

Also, please fill out the required fields in the PR description. For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

/wait

**Comment by RyanTheOptimist** [comment]
> > > It seems there are fast-fail to check if secret is loaded (https://github.com/envoyproxy/envoy/blob/main/source/common/http/http3/conn_pool.cc#L88), so if the secrets will not change to empty again, and the **createQuicNetworkConnection** will not return nullptr.
> > > I wonder if it's necessary to double check in **createQuicNetworkConnection**. If do it, it is necessary to confirm the return value of **createQuicNetworkConnection** out of defense.
> > > Although it's necessary to add a test for fast-fail if secrets not loaded, and I'd like to do it:)
> > 
> > 
> > So to clarify, is the new test testing this existing `return nullptr` (on line 95) or the new code you added? If it's testing the existing code, can we get a test to cover the new code?
> 
> New testing code is for the existing `return nullptr`. And I think the new code which I added is just out of defence, I don't think if the secrets have been loaded before **createQuicNetworkConnection**, it will become empty in that function.

Interesting. I wonder what would happen if you duplicated your new tests, but changed:
`EXPECT_CALL(*config, isReady()).WillRepeatedly(Return(false));`
So that the first call returns true (and hence the old `return nullptr` is not triggered) but *second* call returns false and hence hits the new `return nullptr`. What do you think

**Comment by YaoZengzeng** [comment]
@RyanTheOptimist Updated. Actually we can't manipulate **isReady()**, because it's triggered by secrets change. But we can't precisely make the secret change after fast-fail but before **createQuicNetworkConnection**. 

So I create a mock QuicClientTransportSocketFactory and to achive the goal by changing the return value of method **ssl_context** :)

**Review by RyanTheOptimist — DISMISSED** [review]
This looks good to me. @danzh2010 what do you think?

**Review by danzh2010 — DISMISSED** [review]

File: source/common/http/http3/conn_pool.cc [hunk 1]
```diff
@@ -105,6 +105,12 @@ allocateConnPool(Event::Dispatcher& dispatcher, Random::RandomGenerator& random_
         data.connection_ =
             Quic::createQuicNetworkConnection(h3_pool->quicInfo(), pool->dispatcher(), host_address,
                                               source_address, quic_stat_names, scope);
+        if (data.connection_ == nullptr) {
+          ENVOY_LOG_TO_LOGGER(
+              Envoy::Logger::Registry::getLog(Envoy::Logger::Id::pool), warn,
```
Comment (ggreenway) [inline source/common/http/http3/conn_pool.cc:110]: Is `warn` going to be too much output in the log? I presume if this happens once, it'll probably happen for every connection attempt. Maybe use `ENVOY_LOG_PERIODIC_TO_LOGGER` or `ENVOY_LOG_EVERY_POW_2_TO_LOGGER`?

Is there any failure metric that increases in this case?

**Review by ggreenway — CHANGES_REQUESTED** [review]
/wait-any

File: source/common/http/http3/conn_pool.cc (hunk 1 above)
Comment (YaoZengzeng) [reply source/common/http/http3/conn_pool.cc:111]: @ggreenway updated to **ENVOY_LOG_EVERY_POW_2_TO_LOGGER**. 

As for failure metric, there seems to be no metric to indicate the failure of quic connection creation. From my research, even for HTTP 1/2, there is only a metric  **bind_errors_** which is used to indicate the bind error when failed to create client network connection.

**Review by ggreenway — APPROVED** [review]

**Pull request merged**